content/post/076-async-await-refactor.md

```diff
@@ -0,0 +1,28 @@
+---
+date: 2019-12-10
+url: 2019-12-10-async-await-refactor
+title: The async await refactor
+author: Alan Shaw
+---
+
+> If you're considering a refactor that'll touch ~70 interdependent repos you've come to the right place for some perspective.
+
+We're on the cusp of completing a refactor to the js-ipfs, js-libp2p and js-ipld codebases to use Promises and remove Node.js/pull streams from the code base entirely. We're using `async`/`await` everywhere (i.e. not the `then`/`catch` style of working with promises) and async iterables, allowing users to consume our streaming APIs with `for await (const chunk of iterable)` loops.
+
+First up, if you think you'd like to take on a challenge like this, do the math. If each repo takes a day to refactor and if you work on it 5 days a week then you've automatically used 14 of your 52 weeks in a year. That's like, 4 months, and that's the absolute best case scenario.
+
+In reality, even if you get approval to do it you're unlikely to be able to work on it all day every day. The problem is that the benefits of a refactor like this can't really be seen until everything is done and the whole stack is using the new code. I'll go into the benefits more later, but the point is, **your project needs to keep moving**. New features need to land and those bugs don't squash themselves.
```

Comment:
You say you’ll get into benefits later but don’t here. Edgy about actually starting with this as a “whew, we got to this amazing summit and now we’re going to reflect back on the climb”? I don’t feel like I get a sense of why all this work was relevant or worthwhile in the current structure - I need convincing first that the destination is even interesting/exciting before I’m ready to know how you achieved this long, hard feat...

Reply:
Split into 2 posts to achieve this :)